File: RunLoop线程保活封装类/RunLoop线程保活封装类/ViewController.m
//
//  ViewController.m
//  RunLoop线程保活封装类
//
//  Created by 张旭洋 on 2024/8/26.
//

#import "ViewController.h"
#import "PermenantThread.h"

// 利用RunLoop封装生命周期可控的线程类

@interface ViewController ()

@property (nonatomic, strong)UIButton* button;
@property (nonatomic, strong)PermenantThread* thread;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.thread = [[PermenantThread alloc] init];
    
    self.view.backgroundColor = [UIColor orangeColor];
    
    self.button = [UIButton buttonWithType: UIButtonTypeRoundedRect];
    self.button.frame = CGRectMake(77, 77, 377, 277);
    
    [self.button setTitle: @"停止" forState: UIControlStateNormal];
    [self.button addTarget: self action: @selector(stopAction) forControlEvents: UIControlEventTouchUpInside];
    
    [self.view addSubview: self.button];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.thread executeTask:^{
            NSLog(@"执行任务 - %@", [NSThread currentThread]);
    }];
}

- (void)stopAction {
    [self.thread stop];
}

- (void)dealloc
{
    NSLog(@"%s", __func__);
}


@end

File: NumberDemo/NumberDemo/ViewController.swift
//
//  ViewController.swift
//  NumberDemo
//
//  Created by 张旭洋 on 2023/1/6.
//

import UIKit

class ViewController: UIViewController {
    
    var Number: Int = 0
    var step: Bool = true;
    
    @IBOutlet weak var label: UILabel!
    @IBOutlet weak var button: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        updateUI()
    }
    
    @IBAction func clickButton(_ sender: UIButton) {
        if step {
            Number += 1
        } else {
            Number -= 1
        }
        switch (Number) {
        case 10: step = false; break
        case 0: step = true; break
        default: break
        }
        updateUI()
        
    }
    func updateUI() {
        label.text = "Number:\(Number)"
        let title = !step ? "-1" : "+1"
        button.setTitle(title, for: .normal)
    }
    
}
